File: basic_dsa_ninety_eight.java
/**
 * Basic DSA Problems - Two Sum
 * Find two indices such that their values add up to target
 */
import java.util.*;

public class basic_dsa_ninety_eight {

    // Problem: Given an array of integers nums and target, return indices of the two numbers
    // such that they add up to target. Assume exactly one solution exists.

    /**
     * One pass: store (value -> index) in map; for each num check if (target - num) exists.
     * Time O(n), Space O(n)
     */
    public static int[] twoSum(int[] nums, int target) {
        if (nums == null || nums.length < 2) return new int[] { -1, -1 };
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int need = target - nums[i];
            if (map.containsKey(need)) {
                return new int[] { map.get(need), i };
            }
            map.put(nums[i], i);
        }
        return new int[] { -1, -1 };
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Two Sum");
        System.out.println("=".repeat(60));

        int[] a = { 2, 7, 11, 15 };
        int[] r1 = twoSum(a, 9);
        System.out.println("nums = [2,7,11,15], target = 9 -> " + Arrays.toString(r1));

        int[] b = { 3, 2, 4 };
        int[] r2 = twoSum(b, 6);
        System.out.println("nums = [3,2,4], target = 6 -> " + Arrays.toString(r2));

        int[] c = { 3, 3 };
        int[] r3 = twoSum(c, 6);
        System.out.println("nums = [3,3], target = 6 -> " + Arrays.toString(r3));
    }
}

File: basic_dsa_ninety_eight.cpp
/**
 * Basic DSA - Create Target Array in Given Order (LeetCode 1389)
 * Build target: nums[i] at index[i]; insert at index (shift existing right).
 */
#include <iostream>
#include <vector>

std::vector<int> createTargetArray(std::vector<int>& nums, std::vector<int>& index) {
    std::vector<int> target;
    for (size_t i = 0; i < nums.size(); i++) {
        target.insert(target.begin() + index[i], nums[i]);
    }
    return target;
}

void print(std::vector<int>& v) {
    std::cout << "[";
    for (size_t i = 0; i < v.size(); i++) {
        std::cout << v[i];
        if (i + 1 < v.size()) std::cout << ",";
    }
    std::cout << "]\n";
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Create Target Array in Given Order\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> n1 = {0, 1, 2, 3, 4};
    std::vector<int> i1 = {0, 1, 2, 2, 1};
    auto r1 = createTargetArray(n1, i1);
    std::cout << "nums=[0,1,2,3,4], index=[0,1,2,2,1] -> ";
    print(r1);

    std::vector<int> n2 = {1, 2, 3, 4, 0};
    std::vector<int> i2 = {0, 1, 2, 3, 0};
    auto r2 = createTargetArray(n2, i2);
    std::cout << "nums=[1,2,3,4,0], index=[0,1,2,3,0] -> ";
    print(r2);
}
